client_plugin/drivers/shared/shared_driver.go

```diff
@@ -402,8 +401,161 @@ func (d *VolumeDriver) Path(r volume.Request) volume.Response {
 func (d *VolumeDriver) Mount(r volume.MountRequest) volume.Response {
 	log.WithFields(log.Fields{"name": r.Name}).Info("Mounting volume ")
 
-	log.Errorf("VolumeDriver Mount to be implemented")
-	return volume.Response{Err: ""}
+	// lock the state
+	d.RefCounts.StateMtx.Lock()
+	defer d.RefCounts.StateMtx.Unlock()
+
+	// checked by refcounting thread until refmap initialized
+	// useless after that
+	d.RefCounts.MarkDirty()
+
+	return d.processMount(r)
+}
+
+// processMount -  process a mount request
+func (d *VolumeDriver) processMount(r volume.MountRequest) volume.Response {
+	d.MountIDtoName[r.ID] = r.Name
+
+	// If the volume is already mounted , just increase the refcount.
+	// Note: for new keys, GO maps return zero value, so no need for if_exists.
+	refcnt := d.IncrRefCount(r.Name) // save map traversal
+	log.Debugf("volume name=%s refcnt=%d", r.Name, refcnt)
+	if refcnt > 1 {
+		log.WithFields(
+			log.Fields{"name": r.Name, "refcount": refcnt},
+		).Info("Already mounted, skipping mount. ")
+		return volume.Response{Mountpoint: d.GetMountPoint(r.Name)}
+	}
+
+	if plugin_utils.AlreadyMounted(r.Name, d.MountRoot) {
```

Comment:
"Is there a check needed to verify ... the samba service for this volume and the backing VMDK volume are present. "
The caller of Mount() can only proceed when the state of volume becomes "Mounted". And it's the responsibility of the event handler on one of the manager nodes to start a SMB server, the event handler will fail if they are not able to start the SMB server, and thus the volume will never be "Mounted". And the caller of Mount() will timeout eventually.

"What happens if a mount and a remove are in a race?"
There won't be a mount and remove in a race. This is because all the change of states of the volumes are atomic transactions. A mount can only proceed, if the event handler on one of the manager plugins can change the state from "Ready" to "Mounting". Similarly, a remove can only proceed, if the Remove() function can change the state from "Ready" to "Deleting". If the plugins failed to change the state, they will either wait, or return error, depending on the current state.
So the caller of Mount() will get an error in the case that this volume is being removed (or already removed) by other nodes.